spring-session/src/main/java/org/springframework/session/jdbc/JdbcOperationsSessionRepository.java

```diff
@@ -0,0 +1,439 @@
+	@Override
+	public JdbcSession createSession() {
+		JdbcSession session = new JdbcSession();
+		if (this.defaultMaxInactiveInterval != null) {
+			session.setMaxInactiveIntervalInSeconds(this.defaultMaxInactiveInterval);
+		}
+		return session;
+	}
+
+	@Override
+	public void save(final JdbcSession session) {
+		final byte[] sessionBytes = this.serializer.convert(session.delegate);
+
+		if (session.isNew()) {
+			this.jdbcOperations.update(getQuery(Query.CREATE_SESSION), new PreparedStatementSetter() {
+
+				@Override
+				public void setValues(PreparedStatement ps) throws SQLException {
+					ps.setString(1, session.getId());
+					ps.setLong(2, session.getLastAccessedTime());
+					ps.setString(3, session.getPrincipalName());
+					JdbcOperationsSessionRepository.this.lobHandler.getLobCreator()
+							.setBlobAsBytes(ps, 4, sessionBytes);
+				}
+
+			});
+			session.setNew(false);
+			publishEvent(new SessionCreatedEvent(this, session));
+		}
+		else {
+			this.jdbcOperations.update(getQuery(Query.UPDATE_SESSION), new PreparedStatementSetter() {
+
+				@Override
+				public void setValues(PreparedStatement ps) throws SQLException {
+					ps.setLong(1, session.getLastAccessedTime());
+					ps.setString(2, session.getPrincipalName());
+					JdbcOperationsSessionRepository.this.lobHandler.getLobCreator()
+							.setBlobAsBytes(ps, 3, sessionBytes);
+					ps.setString(4, session.getId());
+				}
+
+			});
+		}
+	}
+
+	@Override
+	public JdbcSession getSession(String id) {
+		MapSession session = null;
+		try {
+			session = this.jdbcOperations.queryForObject(getQuery(Query.GET_SESSION),
+					new Object[] { id }, this.mapper);
+		}
+		catch (EmptyResultDataAccessException ignored) {
+		}
+
+		if (session != null) {
+			if (session.isExpired()) {
+				delete(id, new SessionExpiredEvent(this, id));
+			}
+			else {
+				return new JdbcSession(session);
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public void delete(String id) {
+		delete(id, new SessionDeletedEvent(this, id));
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public Map<String, JdbcSession> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		if (!PRINCIPAL_NAME_INDEX_NAME.equals(indexName)) {
+			return Collections.emptyMap();
+		}
+
+		List<MapSession> sessions = this.jdbcOperations.query(
+				getQuery(Query.LIST_SESSIONS_BY_PRINCIPAL_NAME), new Object[] { indexValue }, this.mapper);
+
+		Map<String, JdbcSession> sessionMap = new HashMap<String, JdbcSession>(sessions.size());
+
+		for (MapSession session : sessions) {
+			sessionMap.put(session.getId(), new JdbcSession(session));
+		}
+
+		return sessionMap;
+	}
+
+	@Scheduled(cron = "0 * * * * *")
+	public void cleanupExpiredSessions() {
+		long now = System.currentTimeMillis();
+		long roundedNow = roundDownMinute(now);
+
+		if (logger.isDebugEnabled()) {
+			logger.debug("Cleaning up sessions expiring at " + new Date(roundedNow));
+		}
+
+		List<String> sessionIds = this.jdbcOperations.queryForList(
```

Comment:
Rather than first querying and then deleting this could just be a delete which would prevent loading all the session ids in memory and reduce overhead on the database.